kivy/ext/mpl/backend_kivy.py

```diff
@@ -0,0 +1,850 @@
+from matplotlib.transforms import Bbox
+from matplotlib.backend_bases import ShowBase
+from matplotlib.mathtext import MathTextParser
+from matplotlib import rcParams
+from hashlib import md5
+from matplotlib import _png
+from matplotlib.font_manager import weight_as_number
+
+try:
+    import kivy
+except ImportError:
+    raise ImportError("this backend requires Kivy to be installed.")
+
+from kivy.app import App
+from kivy.graphics.texture import Texture
+from kivy.graphics import Rectangle
+from kivy.uix.widget import Widget
+from kivy.uix.label import Label
+from kivy.uix.floatlayout import FloatLayout
+from kivy.uix.behaviors import FocusBehavior
+from kivy.uix.actionbar import ActionBar, ActionView, \
+                                ActionButton, ActionToggleButton, \
+                                ActionPrevious, ActionOverflow, ActionSeparator
+from kivy.base import EventLoop
+from kivy.core.text import Label as CoreLabel
+from kivy.graphics import Color, Line
+from kivy.graphics import Rotate, Translate
+from kivy.graphics.tesselator import Tesselator
+from kivy.graphics.context_instructions import PopMatrix, PushMatrix
+from kivy.logger import Logger
+from kivy.graphics import Mesh
+from kivy.resources import resource_find
+from kivy.uix.stencilview import StencilView
+from kivy.core.window import Window
+from kivy.uix.button import Button
+from kivy.uix.boxlayout import BoxLayout
+from kivy.uix.floatlayout import FloatLayout
+from kivy.uix.relativelayout import RelativeLayout
+from kivy.uix.popup import Popup
+from kivy.properties import ObjectProperty
+from kivy.clock import Clock
+
+import numpy as np
+import io
+from _functools import partial
+from math import cos, sin, pi
+
+# try:
+#     kivy.require('1.9.0') # I would need to check which release of
+#     Kivy would be the best suitable.
+# except AttributeError:
+#     raise ImportError(
+#         "kivy version too old -- it must have require_version")
+
+_debug = True
+
+app = None
+
+
+class SaveDialog(FloatLayout):
+    save = ObjectProperty(None)
+    text_input = ObjectProperty(None)
+    cancel = ObjectProperty(None)
+
+
+class MPLKivyApp(App):
+
+    figure = ObjectProperty(None)
+    toolbar = ObjectProperty(None)
+
+    def __init__(self, **kwargs):
+        super(MPLKivyApp, self).__init__(**kwargs)
+        self.figure = kwargs['figure']
+        self.toolbar = kwargs['toolbar']
+
+    def build(self):
+        EventLoop.ensure_window()
+        layout = FloatLayout()
+        layout.clear_widgets()
+        self.figure.size_hint_y = 0.9
+        layout.add_widget(self.figure)
+        self.toolbar.size_hint_y = 0.1
+        layout.add_widget(self.toolbar)
+        return layout
+
+    def on_pause(self):
+        return App.on_pause(self)
+
+    def on_resume(self):
+        App.on_resume(self)
+
+    def on_stop(self):
+        global app
+        app = None
+
+
+def _create_App(fig_canvas, toolbar):
+    global app
+    if app is None:
```

Comment:
Instead of using `app` global use `App.get_running_app()`. Also, if `App.get_running_app()` is not `None` and a app is already running, you should raise an exception here for now.